Make the SCA testable

kursatkobya opened the pull request:
Add SoftwareContainerFactory class to obtain backend container
Add SoftwareContainerAbstractInterface class to be able to create different SC classes for different purposes
Refactor code regarding changes
Add unit tests for agent class

kursatkobya committed:
Make the SCA testable
Add ContainerAbstractInterface class to obtain felible backend
Add SoftwareContainer class to communicate with container backend
Refactor code regarding changes
Add unit tests for agent class

tobsan commented on agent/component-test/softwarecontaineragent_componenttest.cpp:
```diff
@@ -183,7 +185,7 @@ TEST_F(SoftwareContainerAgentTest, ResumeNonSuspendedContainer) {
         ContainerID id = sca->createContainer(valid_config);
 
-        sca->getContainer(id);
+//        sca->getContainer(id);
```
Remove this comment, please.

tobsan commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -88,6 +90,8 @@ SoftwareContainerAgent::SoftwareContainerAgent(Glib::RefPtr<Glib::MainContext> m
                                                 sharedMountsDir,
                                                 shutdownTimeout);
+
+    log_debug() << "Created SCA instance :" ;
```
Please don't use "SCA" in printouts, write agent instead.

tobsan commented on agent/src/softwarecontaineragent.cpp:
```diff
@@ -256,6 +259,12 @@ pid_t SoftwareContainerAgent::execute(ContainerID containerID,
     // Set up a CommandJob for this run in the container
     auto job = container->createCommandJob(cmdLine);
+    if (nullptr == job) {
+        std::string errorMessage("Could not create job instance");
```
`"Could not create command job for " + cmdLine` would be a better errorMessage

tobsan commented on agent/src/softwarecontaineragent.h:
```diff
@@ -79,7 +80,7 @@ class SoftwareContainerAgentError : public SoftwareContainerError
 {
     LOG_DECLARE_CLASS_CONTEXT("SCA", "SoftwareContainerAgent");
-    typedef std::shared_ptr<SoftwareContainer> SoftwareContainerPtr;
+    typedef std::shared_ptr<ContainerAgentInterface> SoftwareContainerPtr;
```
This name is super confusing. Why not SoftwareContainerInterface? It is an interface to SoftwareContainer, isn't it?

kursatkobya replied:
Actually in the first idea the name of that class was ContainerAbstractInterface. Yet there is another ContainerAbstractInterface which connect software container to real back-end impelmentation. So not to create confusion i wanted to make it clear that this is interface used by agent. But may be it can be SoftwareContainerAbstractInterface. What do you think

tobsan replied:
I see. In that case I think `SoftwareContainerAbstractInterface` is better, yes.

tobsan commented on agent/src/softwarecontainerfactory.cpp:
```diff
@@ -0,0 +1,32 @@
+namespace softwarecontainer {
+std::shared_ptr<ContainerAgentInterface>
+SoftwareContainerFactory::createContainer(const ContainerID id,
+                                  std::unique_ptr<const SoftwareContainerConfig> config)
```
indentation

tobsan commented on agent/src/softwarecontainerfactory.h:
```diff
@@ -0,0 +1,46 @@
+namespace softwarecontainer {
+
+/*
+ * This class holds a connection to agent to create containers.
```
Not it doesn't, it is a class that creates containers. Its relation to the agent should be described in the agent, not here.

tobsan commented on agent/src/softwarecontainerfactory.h:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * @file softwarecontainerfactory.h
+ * @brief Contains the softwarecontainer::AgentFactory class
```
This is not the name of the class.

tobsan commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -0,0 +1,213 @@
+{
+public:
+    std::shared_ptr<ContainerAgentInterface> createContainer
+    (const ContainerID id __attribute__((unused)), std::unique_ptr<const SoftwareContainerConfig> config __attribute__((unused)))
```
Indentation, break the line up instead.

kursatkobya replied:
The problem here even one argument passes the 100char rule. But you are right

tobsan commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -0,0 +1,213 @@
+class SoftwareContainerAgentTest: public ::testing::Test
+{
+public:
+    SoftwareContainerAgent *sca;
```
Use a smart pointer instead, then you won't need to delete it.

tobsan commented on libsoftwarecontainer/include/containeragentinterface.h:
```diff
@@ -0,0 +1,130 @@
+
+/**
+ * @file containeragentinterface.h
+ * @brief Contains the softwarecontainer::ContainerAbstractInterface class
```
This name does not match.

tobsan commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -0,0 +1,213 @@
+
+        sca->shutdownContainer(id);
+    });
+}
```
This test is not meaningful imho. Checking that functions we've stubbed does not throw does not help us. We probably want to make sure certain functions gets called when running a certain sequence (and we can stub those I guess).

breakreturn replied:
Agree. The purpose is to have tests for the sequence, i.e. that some calls on the agent results in the expected call(s) to SoftwareContainer.

tobsan reviewed (changes requested):
Looks good generally, some naming and checks that should be changed.

breakreturn commented on agent/src/softwarecontaineragent.h:
```diff
@@ -270,6 +273,11 @@ class SoftwareContainerAgent
     ContainerOptionParser m_optionParser;
+
+    /*
+     * Responsible to provide connection to containers.
```
I think this could be described as the factory. A factory is a common pattern and further details should be in that class. This comment is a bit confusing at the moment.

breakreturn commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -0,0 +1,213 @@
+    {
+        std::unique_ptr<ConfigLoader> loader(new StringConfigLoader(configString));
+        std::unique_ptr<ConfigSource> mainConfig(new MainConfigSource(std::move(loader),
+                                                                      ConfigDefinition::typeMap()));
```
I guess that should be TypeMap() not typeMap(). Did this compile?

kursatkobya replied:
I think TypeMap is a type and typeMap is an object to that type.  And yes it compiles. Ref : `static TypeMap typeMap();`

breakreturn replied:
Oh yeah your'e right, my bad.

kursatkobya committed:
Add proper unit tests for agent
Change the name of ContainerAgentInterface as SoftwareContainerAbstractInterface

breakreturn commented on agent/src/softwarecontainerfactory.cpp:
```diff
@@ -0,0 +1,32 @@
+                                          std::unique_ptr<const SoftwareContainerConfig> config)
+{
+    auto container = std::shared_ptr<SoftwareContainerAbstractInterface>
+                                    (new SoftwareContainer(id, std::move(config)));
```
I think it would be more similar to other parts of the code if the opening parenthesis is on the line above. It's basically the same as when you have a method and need to break the line at the first arg:

longMethodName(
    arg1, arg2);

kursatkobya replied:
it should be like this then : 
```
auto container = std::shared_ptr<SoftwareContainerAbstractInterface>(
                           new SoftwareContainer(id, std::move(config)));
```

breakreturn commented on agent/src/softwarecontainerfactory.h:
```diff
@@ -0,0 +1,46 @@
+
+class SoftwareContainerFactory {
+public:
+    virtual ~SoftwareContainerFactory() {};
```
No need for the ;

breakreturn commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -0,0 +1,213 @@
+    }
+};
+
+*/
```
I guess this can be removed?

breakreturn commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -0,0 +1,213 @@
+{
+    LOG_DECLARE_CLASS_CONTEXT("SCAU", "SoftwareContainerAgent Unit Test");
+public:
+    TestContainerInterface() {};
```
No need for the extra ;

breakreturn commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -0,0 +1,213 @@
+        m_container(testContainerInterface)
+    { }
+
+    std::shared_ptr<SoftwareContainerAbstractInterface> createContainer(const ContainerID id __attribute__((unused)),
```
Will __attribute __() work with clang?

tobsan replied:
I generally don't like the `__attribute__` attributes, I would rather have a compiler warning actually.

tobsan commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -0,0 +1,213 @@
+
+        ASSERT_EQ(testContainerInterface, sca->getContainer(id));
+        auto containers = sca->listContainers();
+        ASSERT_EQ(unsigned(1), containers.size());
```
You could write this as 1u instead.

tobsan reviewed (changes requested):
Mostly minor things. And as I wrote before, I would prefer if we didn't adopt the "SCA" acronym (which goes for commit messages as well).

kursatkobya committed:
Refactor tests and code to imply "no ReturnCode" policy
